### PR TITLE
manual seal is now consensus agnostic

### PR DESCRIPTION
**Pull request by seunlanlege**
This PR introduces the trait `ConsensusDataProvider` which manual-seal uses to author blocks valid for any runtime.

### CODE REVIEW

File: client/consensus/manual-seal/src/consensus/babe.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+			)
+			.ok_or_else(|| {
+				log::info!(target: "babe", "create_digest: no viable_epoch :(");
+				sp_consensus::Error::InvalidAuthoritiesSet
+			})?;
+
+		if epoch.as_ref().end_slot()  == slot_number {
+			// we want to create a new block here.
+			log::info!(target: "manual-seal", "\n\nthis is the end of the epoch\n\n")
+		}
```
Comment (seunlanlege) [inline client/consensus/manual-seal/src/consensus/babe.rs:127]: oops

File: client/consensus/manual-seal/Cargo.toml [hunk 2]
```diff
@@ -22,20 +22,28 @@ parking_lot = "0.10.0"
...
+sc-transaction-pool = { path = "../../transaction-pool", version = "2.0.0-rc5" }
+sp-blockchain = { path = "../../../primitives/blockchain", version = "2.0.0-rc5" }
+sp-consensus = { package = "sp-consensus", path = "../../../primitives/consensus/common", version = "0.8.0-rc5" }
+sp-inherents = { path = "../../../primitives/inherents", version = "2.0.0-rc5" }
+sp-runtime = {  path = "../../../primitives/runtime", version = "2.0.0-rc5" }
+sp-core = {  path = "../../../primitives/core", version = "2.0.0-rc5" }
+sp-api = {  path = "../../../primitives/api", version = "2.0.0-rc5" }
+sp-transaction-pool = { path = "../../../primitives/transaction-pool", version = "2.0.0-rc5" }
+sp-timestamp = { path = "../../../primitives/timestamp", version = "2.0.0-rc6" }
+tokio = { version = "0.2", features = ["rt-core", "macros", "time"] }
```
Comment (tomusdrw) [inline client/consensus/manual-seal/Cargo.toml:40]: Can you double check that we really need all of these deps? Where is `tokio` used for instance?

File: client/consensus/manual-seal/src/consensus/babe.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+use sp_runtime::{
+	traits::{DigestItemFor, DigestFor, Block as BlockT, Header as _},
+	generic::Digest,
+};
+use sc_client_api::AuxStore;
+use sp_api::{ProvideRuntimeApi, TransactionFor};
+use sc_consensus_epochs::{SharedEpochChanges, Epoch as _, descendent_query};
+use sp_blockchain::{HeaderBackend, HeaderMetadata};
+use sp_consensus::BlockImportParams;
+use sp_timestamp::{InherentType, InherentError, INHERENT_IDENTIFIER};
```
Comment (tomusdrw) [inline client/consensus/manual-seal/src/consensus/babe.rs:47]: Would be really good to keep the imports sorted.

**Review by tomusdrw — APPROVED** [review]
Looks good! For the context: This allows "forging" fake babe blocks, when running a custom node with a `BabeVerifier` that skips signature checks. This is super useful for testing stuff by forking an existing network with the entire database (blockchain, state, runtime), so for instance it make it possible to test if the network still works after a runtime upgrade is enacted.

**Comment by seunlanlege** [comment]
bot merge

**Pull request merged**